pkg/router/crl/crl.go

```diff
@@ -0,0 +1,441 @@
+		}
+	}
+	log.Info("failed to get valid CRL after trying all distribution points")
+	return nil, kerrors.NewAggregate(errs)
+}
+
+// getHTTPCRL gets a certificate revocation list using the provided HTTP URL.
+func getHTTPCRL(url string) (*pkix.CertificateList, error) {
+	resp, err := http.Get(url)
+	if err != nil {
+		return nil, fmt.Errorf("http.Get failed: %w", err)
+	}
+	defer resp.Body.Close()
+	bytes, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, fmt.Errorf("error reading response: %w", err)
+	}
+	crl, err := x509.ParseCRL(bytes)
+	if err != nil {
+		return nil, fmt.Errorf("error parsing response: %w", err)
+	}
+	return crl, nil
+}
+
+// reapStaleDirectories deletes any subdirectories of mtlsBaseDirectory that are no longer necessary
+func reapStaleDirectories() error {
+	log.V(4).Info("Cleaning up stale mtls files...")
+	// list all entries in the base directory.
+	dirEntries, err := os.ReadDir(mtlsBaseDirectory)
+	if err != nil {
+		return err
+	}
+	// The base directory should only contain 2 entries: the latest mTLS files, and the "latest" symlink. If it only
+	// contains 2 items, assume it's those two, so nothing needs to be done.
+	if len(dirEntries) <= 2 {
+		log.V(2).Info("No stale directories")
+		return nil
+	}
+	// Get the name of the directory that mtlsLatestSymlink points to so we don't delete it. If os.Readlink() returns an
+	// error, it'll be because mtlsLatestSymlink doesn't exist. That shouldn't happen, but if it does, we can't
+	// determine what we need to keep, so just exit and hope it's fixed in a future run.
+	inUseDirName, err := os.Readlink(mtlsLatestSymlink)
+	if err != nil {
+		log.V(1).Info(fmt.Sprintf("Couldn't find %s. Can't determine which files to clean", mtlsLatestSymlink))
+		return nil
+	}
+	// Walk through all entries in mtlsBaseDirectory. Only 2 things need to stay: mtlsLatestSymlink, and the directory
+	// that mtlsLatestSymlink points to.  Everything else can be deleted.
+	var errs []error
+	for _, dirEntry := range dirEntries {
+		entryName := dirEntry.Name()
+		// entryName is the bare file/directory name, but mtlsLatestSymlink and inUseDirName are fully-qualified paths.
+		// Since we know that all the dir entries are the files/directories in mtlsBaseDirectory, we can prepend that to
+		// entryName to get the full name to compare against.
+		fullPath := filepath.Join(mtlsBaseDirectory, entryName)
+		log.V(4).Info(fmt.Sprintf("Checking %s (%s)", entryName, fullPath))
+		// Leave mtlsLatestSymlink and the current directory pointed to by mtlsLatestSymlink alone.
+		if fullPath == mtlsLatestSymlink || fullPath == inUseDirName {
+			continue
+		}
+		if err := os.RemoveAll(fullPath); err != nil {
+			log.Error(err, "Failed to remove stale directory", "directory", fullPath)
+			errs = append(errs, err)
+		}
+	}
+	log.V(4).Info("Cleanup done")
+	if len(errs) != 0 {
+		return kerrors.NewAggregate(errs)
```

Comment:
It usually isn't necessary to log _and_ return an error.  In this case, the caller logs any error that it gets, so the router will end up logging each error twice.  Is there any reason not simply to log errors in `reapStaleDirectories` and eliminate its return value?

Reply:
I think it makes sense to eliminate the return value, and log any errors that occur instead. `updateCRLFile` already only calls `reapStaleDirectories` in a `defer` call, so returning an error and logging it in the `defer` doesn't add anything. I missed this in my latest change, but I'll update it shortly.